### ISSUE TITLE
Some questions about compilers

### ISSUE THREAD

**Issue by izZCrown** [monkbai/DNN-decompiler#2]
Hello. 
I recently tried to fully reproduce your experiment, including compiling the model and disassembling it using ida. I noticed that the ida script you provided uses a ". out" file as input to ida. Well, to be honest, I know relatively little about this aspect. When I compile the model using Glow, I generate four files: ". o,. h,. bin, and. txt", which are mentioned in the official Glow documentation. 

<img width="840" alt="image" src="https://user-images.githubusercontent.com/44794620/226550167-6c4195ca-28d4-4265-ab9e-96b4c5532f94.png">

So I want to ask how this ". out" file was generated, which seems like a silly question.  
In addition, I would like to ask what file is inputted into ida after being compiled by TVM?

**Comment by monkbai** [comment]
Hi, thank you for the question. 

After compilation, we need to provide a "main" function that takes input and calls DNN operators implemented in the object file. Specifically, ".out" is a standalone executable (elf) with the object file and model weights inside. I have uploaded an example [here](https://github.com/monkbai/DNN-decompiler/blob/master/utils/main.c) for your reference. 

You can compile the main.c file with a traditional C compiler (e.g., gcc) like ```gcc ./main.c ./test.h ./test.o -o ./test.out```. Glow officially provide a slightly more complex example that takes an image as input ([main.cpp](https://github.com/pytorch/glow/blob/master/examples/bundles/resnet50/main.cpp)). We use such standalone executables as inputs to ida.

As for TVM, we simply follow the officially provided compilation [pipeline](https://github.com/apache/tvm/tree/main/apps/bundle_deploy), which outputs a statically linked x64 standalone executable.

**Comment by izZCrown** [comment]
@monkbai Thank you very much for your detailed response, I think this is exactly what I need.

**Comment by izZCrown** [comment]
Hi, @monkbai 
Sorry to bother you again. I have another question to ask. I noticed that the ". out" file uses ". bin" as input to output classification results, right. 

I noticed that Glow officially provided a "png2bin" tool, but I received an error during the process of using it. I tried to write my own script but the output classification results were incorrect.

So I would like to ask if you also use this tool? Or did you write your own script?

**Comment by monkbai** [comment]
Hi @izZCrown , I reuse the [script](https://github.com/apache/tvm/blob/main/apps/bundle_deploy/build_model.py#L129) provided in the official TVM [budle_deploy](https://github.com/apache/tvm/tree/main/apps/bundle_deploy) example to read an image and produce the binary file.

**Comment by izZCrown** [comment]
> Hi @izZCrown , I reuse the [script](https://github.com/apache/tvm/blob/main/apps/bundle_deploy/build_model.py#L129) provided in the official TVM [budle_deploy](https://github.com/apache/tvm/tree/main/apps/bundle_deploy) example to read an image and produce the binary file.

Well, it seems that I still need to carefully read the official document. Thank you for your reply, which has been very helpful to me.

**Comment by izZCrown** [comment]
Hello, @monkbai 
I recently encountered two problems and would like to ask you for advice.
- When I use the data in BTD-data to decompile, I can get a ```func_call.log``` with data. When I tried to use my own ```resnet18.out``` and ```resnet18_funcs```, I noticed that this command is executed normally. But there is nothing in ```func_call.log``` except ```#eof```. I don't know what's wrong, so I want to ask if you have encountered the same problem?

<img width="894" alt="image" src="https://user-images.githubusercontent.com/44794620/229670759-7fece8be-4cff-4ac1-adc9-aa8c35d66f79.png">

- I noticed that almost every model has a ```config.json``` file in its own folder. So I'm wondering what does it do? And how do I get this file if I use my own model?

<img width="371" alt="image" src="https://user-images.githubusercontent.com/44794620/229670979-f3b9513f-0d5c-417b-8303-58c19f4c7771.png">

Looking forward to your reply very much.

**Comment by monkbai** [comment]
Hi, thanks for asking. For the first question, could you please debug the arguments passed to func_call_rdi_rsi? This pin tool, as described by its name, will record the `rdi`, `rsi`, and some other registers' values at the entry of each function. If the address list provide to it is empty, then there will be no values logged.

For the second question, BTD currently leverages some assembly function prototypes of DNN operators (as we discussed in Appendix B in [our paper](https://arxiv.org/pdf/2210.01075.pdf)). Such prototype information is stored in `config.json`, which is basically model-agnostic and determined by the DL compiler. So you can reuse the config file if you are using the same DL compiler (i.e., using config stored in ***_tvm for TVM and config in ***_glow for Glow). You can also simply reverse some DNN executables to collect such information for other operators/compilers.

**Comment by izZCrown** [comment]
> config

Hello, thanks for the reply. For the first question, if I understand correctly, the address list in ```addrs_rdi_rsi_tmp.log``` should be provided to ```fun_call_rdi_rsi, right```? If so, yes, this address list exists, and it is indeed the entry address of each function.

**Comment by izZCrown** [comment]
> Hi, thanks for asking. For the first question, could you please debug the arguments passed to func_call_rdi_rsi? This pin tool, as described by its name, will record the `rdi`, `rsi`, and some other registers' values at the entry of each function. If the address list provide to it is empty, then there will be no values logged.
> 
> For the second question, BTD currently leverages some assembly function prototypes of DNN operators (as we discussed in Appendix B in [our paper](https://arxiv.org/pdf/2210.01075.pdf)). Such prototype information is stored in `config.json`, which is basically model-agnostic and determined by the DL compiler. So you can reuse the config file if you are using the same DL compiler (i.e., using config stored in ***_tvm for TVM and config in ***_glow for Glow). You can also simply reverse some DNN executables to collect such information for other operators/compilers.

Hello, thanks for the reply. For the first question, if I understand correctly, the address list in ```addrs_rdi_rsi_tmp.log``` should be provided to ```fun_call_rdi_rsi```, right? If so, yes, this address list exists, and it is indeed the entry address of each function.

**Comment by monkbai** [comment]
Yes, you are correct. Addresses will be stored in `addrs_rdi_rsi_tmp.log`, the pin tool is invoked afterward. In this case, could you please debug the corresponding [pin tool](https://github.com/monkbai/DNN-decompiler/blob/691b74baf40893e5051590b0836cd0dc17c06e1e/MyPinTool/FunCallRdiRsi.cpp#L28) to check if it works as expected?

**Comment by izZCrown** [comment]
> Yes, you are correct. Addresses will be stored in `addrs_rdi_rsi_tmp.log`, the pin tool is invoked afterward. In this case, could you please debug the corresponding [pin tool](https://github.com/monkbai/DNN-decompiler/blob/691b74baf40893e5051590b0836cd0dc17c06e1e/MyPinTool/FunCallRdiRsi.cpp#L28) to check if it works as expected?

Hello, thanks for the reply. According to your reminder, I debugged pintool and finally found that it was my model's problem. 

Thanks for your help.

**Comment by monkbai** [comment]
Basically, the two executables are the same, except the `strip.out` one is stripped. We use IDA to disassemble the executable with debug information only to get the ground truth of function labels as training/testing data for the operator recovery model described in Sec. 4.1 in our paper. BTD takes stripped executable as input, and disassembles it with IDA. Since we only need (1) the range of functions and (2) the disassembly code, IDA can perform this task very well.

The hint only applies when padding is involved in the convolutional layers. As shown in the example below, if the randomly picked element is close to the "edges" of the output matrix, the calculation involves some padding zeros (gray boxes), which makes the symbolic expressions incomplete. However, according to your description that there is no expression in the `mem_exp.log`, I suspect that there might be unexpected errors in the symbolic execution process that have not been caught. You may debug the process and see if the sliced log is generated correctly and if the symbolic execution is finished accordingly.
![example](https://user-images.githubusercontent.com/38723851/231965518-12e644c2-7477-4d58-ac8e-95b6f3153524.png)

**Comment by izZCrown** [comment]
> Basically, the two executables are the same, except the `strip.out` one is stripped. We use IDA to disassemble the executable with debug information only to get the ground truth of function labels as training/testing data for the operator recovery model described in Sec. 4.1 in our paper. BTD takes stripped executable as input, and disassembles it with IDA. Since we only need (1) the range of functions and (2) the disassembly code, IDA can perform this task very well.
> 
> The hint only applies when padding is involved in the convolutional layers. As shown in the example below, if the randomly picked element is close to the "edges" of the output matrix, the calculation involves some padding zeros (gray boxes), which makes the symbolic expressions incomplete. However, according to your description that there is no expression in the `mem_exp.log`, I suspect that there might be unexpected errors in the symbolic execution process that have not been caught. You may debug the process and see if the sliced log is generated correctly and if the symbolic execution is finished accordingly. ![example](https://user-images.githubusercontent.com/38723851/231965518-12e644c2-7477-4d58-ac8e-95b6f3153524.png)

Okay, thank you for your reply. I will try debugging it again.
